pkg/ccl/changefeedccl/changefeed_processors.go

```diff
@@ -1693,21 +1690,28 @@ func (f *schemaChangeFrontier) SpanFrontier() *span.Frontier {
 }
 
 // getCheckpointSpans returns as many spans that should be checkpointed (are
-// above the highwater mark) as can fit in maxBytes.
-func (f *schemaChangeFrontier) getCheckpointSpans(maxBytes int64) (checkpoint []roachpb.Span) {
+// above the highwater mark) as can fit in maxBytes, along with the earliest
+// timestamp of the checkpointed spans.
+func (f *schemaChangeFrontier) getCheckpointSpans(
+	maxBytes int64,
+) (spans []roachpb.Span, timestamp hlc.Timestamp) {
 	var used int64
 	frontier := f.frontierTimestamp()
+	checkpointFrontier := hlc.Timestamp{}
 	f.Entries(func(s roachpb.Span, ts hlc.Timestamp) span.OpResult {
 		if frontier.Less(ts) {
 			used += int64(len(s.Key)) + int64(len(s.EndKey))
 			if used > maxBytes {
 				return span.StopMatch
 			}
-			checkpoint = append(checkpoint, s)
+			spans = append(spans, s)
+			if checkpointFrontier.IsEmpty() || ts.Less(checkpointFrontier) {
+				checkpointFrontier = ts
+			}
 		}
 		return span.ContinueMatch
 	})
-	return checkpoint
+	return spans, checkpointFrontier
```

Comment:
Logically, this method is okay -- does exactly what it should.  However, I think we should do a bit more to make checkpoints more efficient.  This can be done here, but probably better do it at the call site if we are checkpointing due to lagging spans.  The problem is as follows: during backfills (or initial scans) *all* emitted resolved spans have the same timestamp.  So, when you add them to frontier, contiguous spans get merged.
That might not be the case here -- as a matter of fact, it won't.  Instead, we will have whole bunch of spans that have slightly different timestamps -- and so, they won't be merged.

Now, once you have figured out checkpointFrontier -- all of these spans have exactly the same timestamp.
So, you can construct a temporary SpanGroup object purely for the purpose of getting adjacent spans merged.
And you use that to checkpoint.

Looking over this method, I actually think we should revert the changes you have made here, and introduce
a standalone method something like `getLaggingCheckpointSpans`.  Basically, I don't want to do any filtering due to the size limits until we have built a SpanGroup, got everything neatly collapsed, and then limit collapsed set based on the size.